src/Kinto.elm

```diff
@@ -0,0 +1,316 @@
+module Kinto exposing (..)
+
+import Base64
+import Http
+import Json.Decode as Json exposing ((:=))
+import String
+import Task exposing (Task)
+
+
+type alias Url =
+    String
+
+
+type alias Verb =
+    String
+
+
+
+-- Auth
+
+
+type alias Username =
+    String
+
+
+type alias Password =
+    String
+
+
+type alias Token =
+    String
+
+
+type Auth
+    = NoAuth
+    | Basic Username Password
+    | Bearer Token
+
+
+
+-- Kinto types
+
+
+type alias BucketName =
+    String
+
+
+type alias CollectionName =
+    String
+
+
+type alias RecordId =
+    String
+
+
+type Endpoint
+    = RootEndpoint
+    | BucketEndpoint (Maybe BucketName)
+    | CollectionEndpoint BucketName (Maybe CollectionName)
+    | RecordEndpoint BucketName CollectionName (Maybe RecordId)
+
+
+type alias Config =
+    { baseUrl : String
+    , headers : List ( String, String )
+    }
+
+
+
+-- Kinto errors
+
+
+type alias ErrorRecord =
+    { errno : Int
+    , message : String
+    , code : Int
+    , error : String
+    }
+
+
+type alias StatusCode =
+    Int
+
+
+type alias StatusMsg =
+    String
+
+
+type Error
+    = ConfigError String
+    | ServerError StatusCode StatusMsg String
+    | KintoError StatusCode StatusMsg ErrorRecord
+    | NetworkError Http.RawError
+
+
+
+-- Making requests
+
+
+endpointUrl : Config -> Endpoint -> Url
+endpointUrl config endpoint =
+    let
+        baseUrl =
+            if String.endsWith "/" config.baseUrl then
+                String.dropRight 1 config.baseUrl
+            else
+                config.baseUrl
+
+        joinUrl =
+            String.join "/"
+    in
+        case endpoint of
+            RootEndpoint ->
+                -- Otherwise Kinto returns a 307
+                -- See https://github.com/Kinto/kinto/issues/852
+                baseUrl ++ "/"
+
+            BucketEndpoint Nothing ->
+                joinUrl [ baseUrl, "buckets" ]
+
+            BucketEndpoint (Just bucketName) ->
+                joinUrl [ baseUrl, "buckets", bucketName ]
```

Comment:
Why have just one `BucketEndpoint` constructor with a `Maybe` for these two cases? It seems like these two things are really two different endpoints, so I'd expect them to be two different constructors. (Same for all the subsequent endpoints.) I guess they're really equivalent so it doesn't matter that much, especially since this is an implementation detail that isn't exposed.